uaclient/log.py

```diff
@@ -0,0 +1,59 @@
+import json
+import logging
+import re
+from collections import OrderedDict
+from typing import Any, Dict, List
+
+
+class JsonArraryFormatter(logging.Formatter):
+    default_time_format = "%Y-%m-%dT%H:%M:%S%z"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if not self.datefmt:
+            self.datefmt = self.default_time_format
+        self._required_fields = self.get_required_fields()
+
+    def get_required_fields(self) -> List[str]:
+        # We only support the % style formatting...currently
+        if isinstance(self._style, logging.PercentStyle):
+            formatter_style_pattern = re.compile(r"%\((.+?)\)", re.IGNORECASE)
+        else:
+            raise ValueError("Invalid format: %s" % self._fmt)
+
+        if self._fmt:
+            return formatter_style_pattern.findall(self._fmt)
+        else:
+            return []
+
+    def update_local_log_records(
```

Comment:
I don't think this needs to be a separate function. Let's just do it down in `format`

Reply:
You are answering all questions I had :-) :-)